### PR TITLE
proxy deployer contract

### PR DESCRIPTION
**Pull request by psorinionut**

### CODE REVIEW

**Review by CostinCarabas — DISMISSED** [review]

File: contracts/proxy-deployer/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+version = "0.0.0"
+authors = ["MultiversX <contact@multiversx.com>"]
+edition = "2021"
+publish = false
+
+[lib]
+path = "src/lib.rs"
+
+[dependencies.multiversx-sc]
+version = "=0.43.4"
```
Comment (CostinCarabas) [inline contracts/proxy-deployer/Cargo.toml:12]: Please upgrade framework version. All the contracts have the framework version 0.44.

**Review by CostinCarabas — CHANGES_REQUESTED** [review]

**Review by CostinCarabas — CHANGES_REQUESTED** [review]

File: contracts/proxy-deployer/Cargo.toml (hunk 1 above)
Comment (psorinionut) [reply contracts/proxy-deployer/Cargo.toml]: Done.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    fn contract_upgrade(
+        &self,
+        contract_address: ManagedAddress,
+        template_address: ManagedAddress,
+        args: MultiValueEncoded<ManagedBuffer>,
+    ) {
+        self.require_not_paused();
+        self.check_caller_can_call_endpoint(&contract_address);
+        require!(
+            self.blockchain().is_smart_contract(&template_address),
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:56]: Added.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            "Template address is not a SC"
+        );
+
+        let arguments = args.to_arg_buffer();
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs]: Updated.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        self.check_caller_can_call_endpoint(&contract_address);
+
+        self.send()
+            .contract_call::<()>(contract_address, function_name)
+            .with_gas_limit(self.blockchain().get_gas_left())
+            .with_raw_arguments(args.to_arg_buffer())
+            .execute_on_dest_context()
+    }
+
+    #[only_owner]
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:90]: Added.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    #[endpoint(upgradeContractsByTemplate)]
+    fn upgrade_contracts_by_template(
+        &self,
+        gas_per_action: u64,
+        opt_template_address: OptionalValue<ManagedAddress>,
+        opt_args: OptionalValue<MultiValueEncoded<ManagedBuffer>>,
+    ) -> bool {
+        let mut ongoing_upgrade_operation =
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:99]: I changed the code a bit, to always allow the owner to interact with the contracts, even if the proxy deployer is paused.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+            self.get_ongoing_operation(opt_template_address, opt_args);
+        while self.blockchain().get_gas_left() >= gas_per_action + DEFAULT_GAS_FOR_SAVE
+            && !ongoing_upgrade_operation.contracts_remaining.is_empty()
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:102]: That if statement is outside the while statement. The idea is to repeat the process until there is no gas left or all contracts were processed. After we exit the while block, we check which case was first and update the storage accordingly.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    fn get_ongoing_operation(
+        &self,
+        opt_template_address: OptionalValue<ManagedAddress>,
+        opt_args: OptionalValue<MultiValueEncoded<ManagedBuffer>>,
+    ) -> OngoingUpgradeOperation<Self::Api> {
+        if opt_template_address.is_none() {
+            require!(
+                !self.ongoing_upgrade_operation().is_empty(),
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs]: Done.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            "Template address is not a SC"
+        );
+        let contracts_by_template = self
+            .deployed_contracts_list_by_template(&template_address)
+            .get();
+        require!(
+            !contracts_by_template.is_empty(),
+            "No contracts deployed with this template"
+        );
+        let args = opt_args
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs]: Updated.

**Review by CostinCarabas — DISMISSED** [review]

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            self.blockchain().is_smart_contract(&template_address),
+            "Template address is not a SC"
+        );
+
+        self.send_raw().upgrade_from_source_contract(
+            &contract_address,
+            self.blockchain().get_gas_left(),
+            &BigUint::zero(),
+            &template_address,
+            CodeMetadata::DEFAULT,
```
Comment (CostinCarabas) [inline contracts/proxy-deployer/src/contract_interactions.rs:73]: Is there a way where you can keep the previous CodeMetadata? Maybe after upgrade, the owner doesn't want to change them to default.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+        require!(
+            self.blockchain().is_smart_contract(&contract_address),
+            "Contract address is not a SC"
+        );
+
+        self.send()
+            .contract_call::<()>(contract_address, function_name)
+            .with_gas_limit(self.blockchain().get_gas_left())
+            .with_raw_arguments(args.to_arg_buffer())
+            .execute_on_dest_context()
```
Comment (CostinCarabas) [inline contracts/proxy-deployer/src/contract_interactions.rs:95]: Are there going to be only intra-shard sync calls?
Will there going to be a deployer on each shard?

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            let contract_address = ongoing_upgrade_operation
+                .contracts_remaining
+                .get(0)
+                .clone_value();
+            self.send_raw().upgrade_from_source_contract(
+                &contract_address,
+                gas_per_action,
+                &BigUint::zero(),
+                &ongoing_upgrade_operation.template_address,
+                CodeMetadata::DEFAULT,
```
Comment (CostinCarabas) [inline contracts/proxy-deployer/src/contract_interactions.rs:123]: Same here. You are changing the CodeMetadata.

**Review by CostinCarabas — APPROVED** [review]

File: contracts/proxy-deployer/src/contract_interactions.rs (hunk 9 above)
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:73]: Yes, we plan to have it working this way (keeping the previous CodeMetadata). But we need to wait for the upgrade + a new function in the framework.

File: contracts/proxy-deployer/src/contract_interactions.rs (hunk 10 above)
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:95]: When you deploy a new contract, it is deployed in the same shard as the owner. So right now, we'll have only intra shard calls. Not sure if in the future we will change this.

File: contracts/proxy-deployer/src/contract_interactions.rs (hunk 11 above)
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:123]: The same answer as above.

**Review by CostinCarabas — DISMISSED** [review]

**Review by mihaicalinluca — APPROVED** [review]

**Review by CostinCarabas — APPROVED** [review]

**Pull request merged**